src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -285,24 +286,23 @@ class Ledger(vm: VM) extends Logger {
     * @param worldStateProxy
     * @return
     */
-  private[ledger] def updateSenderAccountBeforeExecution(stx: SignedTransaction, worldStateProxy: InMemoryWorldStateProxy):
-  (InMemoryWorldStateProxy, Address) = {
-    val (world, recipient) = stx.tx.receivingAddress match {
+  private[ledger] def updateSenderAccountBeforeExecution(stx: SignedTransaction, worldStateProxy: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+    val senderAddress = stx.senderAddress
+    val account = worldStateProxy.getGuaranteedAccount(senderAddress)
+    worldStateProxy.saveAccount(senderAddress, account.increaseBalance(-calculateUpfrontGas(stx.tx)).increaseNonce)
+  }
+
+  private[ledger] def prepareTomato(stx: SignedTransaction, worldStateProxy: InMemoryWorldStateProxy) = stx.tx.receivingAddress match {
```

Comment:
Maybe the tomato should be the `ProgramContext`? And all (most of) former `ProgramContext#apply# should go there?